### DISCUSSION TITLE
Stack Navigator in capacitor

### DISCUSSION THREAD

**Discussion by ValGeorgiev** [ionic-team/capacitor#5695] [Q&A]
Hey all! First of all thanks for the great library. We are looking forward to integrate our app with capacitor.js. One of the features we are looking to add is the native swipe transitions between routes. Here is an example: 

https://user-images.githubusercontent.com/12056467/174078467-7d914013-2896-4bb8-920f-66ffbab835e8.mov

Is this possible with Capacitor? We are using react-router v6 for routing. Any suggestions are much appreciated.

**Comment by mlynch** [comment]
@ValGeorgiev this is absolutely possible, but you'll need a UI kit that handles this. This is actually one of the core benefits to using Ionic Framework since it has native-quality navigation built in and integrates with React Router when using React.

You could use Ionic Framework's React components and only use the nav/router integration if you want to use your own components elsewhere. See this Tailwind + Next.js + Capacitor + Ionic Framework starter I built a while back that does just that: https://github.com/mlynch/nextjs-tailwind-ionic-capacitor-starter

**Reply by ValGeorgiev** [reply]
Hey @mlynch thanks for your quick response. I cloned your repository and tried it, but I am looking for more smooth back navigation. I was able to experience the following. (tested on real device as well)

https://user-images.githubusercontent.com/12056467/174147652-506c1ddb-a1c8-4c4a-a166-aac82b053630.mov

 Do we have the option to see both screens while transitioning with ionic router? Also it looks like ionic router depends on react-router-dom@5 not the latest 6. Do we have a plan to address this in the future? Sorry for the direct questions. Thanks again

**Reply by mlynch** [reply]
That's probably just a bug with the way the routes are specified in this demo. You shouldn't be able to swipe to go back from that tab. You'll see traditional smooth swipe to go back if you have a normal `push` navigation. I'll look into fixing it.

**Reply by ollyde** [reply]
@mlynch I'm sorry but Ionic does not have a native quality navigation utility. We tried using it on some of our test devices, it was buggy and janky. For example the "Nothing" phone it's completely broken.

You guys really need a native stack inside Cap :-)

**Reply by mlynch** [reply]
Yes it does. It's being used successfully in many hundreds of thousands of apps. Sometimes it's not configured correctly though and that can cause issues, or developers are doing too much work at the start of the transition and not setting up lifecycle events correctly.

Native stack navigation is a possible addition to capacitor, if someone built that we would share it to teams that need it.

**Reply by Elvincth** [reply]
It would be great if native stack navigation could be supported, as currently iOS WebView doesn't allow for 120Hz transitions, a bit laggy for CSS transitions.

**Comment by brandonxiang** [comment]
I think maybe a native stack navigator will be better. It is possible that capacitor provide a plugin to support the native app stack navigator?

**Comment by ollyde** [comment]
@ValGeorgiev  It's one of the main reasons I think capacitor isn't a good library for serious mobile/web developers. 
We were forced to use Flutter (for Web and mobile) because of the poor quality transitions in Capacitor. 

@mlynch Copy Flutter with a simple native navigation stack and actions like ```await Nav.pop()``` or ```await Nav.goToUrl([url, parms])```

Making the library create it's own navigation effects isn't producing good results.. For example we'd love to use Svelte, but it's not possible right now. Just make the navigation a native stack.

**Reply by mlynch** [reply]
Capacitor doesn't handle any UI for you. If you want native mobile UI, use it with Ionic Framework. 

I would love to bring more of those primitives to Capacitor proper, but it's tricky because web devs often want to use their existing routing/etc. which means the UI layer has to hook into each framework. This is why Ionic Framework only supports Angular, React, and Vue.

You can roll your own navigation and transitions outside of Ionic Framework, and we've seen teams have a lot of success doing that with Tailwind and projects like Framer Motion, but we don't have any official offering for those.

Capacitor is absolutely a great library for serious web and mobile development, but we need to do more to bring that to users of Svelte and other libraries where using Ionic Framework isn't an option.

**Reply by ollyde** [reply]
@mlynch thanks for the reply. As I mentioned before the 'emulation' transitions never work properly, they always end with poor results, bugs, 'unnatural feeling' and many more issues. 

We've been holding off PWAs (and many others for over a decade now) because of this 1 core issue. If you had native transitions cap would be perfect! We'd ditch RN, Flutter and other frameworks for it tomorrow.

As a user I expect to be able to swipe back and see forward transitions in my apps, use back buttons properly, get transitions for it. Many Android manufacturers completely screw up web navigation.

For example, I have a "Nothing Phone (brand)" that my QA uses. Capacitor + Ionic is completely unusable, why? The "Nothing phone" has it's own navigation swipe back for Android, no back buttons using it's own native stack swipe back area. This is just one case..

**Reply by ollyde** [reply]
@mlynch you can also have multiple web-views with a navigation controller, I never mentioned anything about rending UI, I know that is web-based. This is actually a fairly simple fix looking at the capacitor code.

**Comment by vojto** [comment]
It's worth noting that a few apps are now taking the hybrid approach:

1. Screenshot the current web view
2. Move the webview to the target view
3. Perform a native transition from one to another

[37signals Turbo](https://github.com/hotwired/turbo-ios) framework has been using that trick for a few years. More recently, [Height](https://twitter.com/michaelvillar/status/1589659903076859904) app implemented the same technique. 

If someone ends up creating a Capacitor plugin for this, I would love to test it out!

**Reply by ollyde** [reply]
When swiping back it's broken though. We need multiple webviews.
